share/transformers/gov_nsfawards.py

```diff
@@ -54,54 +67,178 @@ class AffiliatedAgent(Parser):
+        awardee_zip_code = tools.Try(ctx.awardeeZipCode)
 
 
 class IsAffiliatedWith(Parser):
     related = tools.Delegate(AffiliatedAgent, ctx)
 
 
-class ContributorAgent(Parser):
+class AgentIdentifier(Parser):
+    uri = ctx
+
+
+class POContributorAgent(Parser):
+    schema = 'Person'
+
+    name = ctx.poName
+
+    identifiers = tools.Map(
+        tools.Delegate(AgentIdentifier),
+        tools.Try(
+            tools.IRI(ctx.poEmail),
+            exceptions=(ValueError,)
+        )
+    )
+
+    related_agents = tools.Map(
+        tools.Delegate(IsAffiliatedWith),
+        tools.Filter(lambda x: 'awardeeName' in x, ctx)
+    )
+
+    class Extra:
+        po_name = tools.Try(ctx.poName)
+        po_email = tools.Try(ctx.poEmail)
+
+
+class POContributorRelation(Parser):
+    schema = 'Contributor'
+
+    agent = tools.Delegate(POContributorAgent, ctx)
+    cited_as = ctx.poName
+
+
+class PIContributorAgent(Parser):
     schema = 'Person'
 
     family_name = ctx.piLastName
     given_name = ctx.piFirstName
+    additional_name = tools.Try(ctx.piMiddeInitial)
 
-    related_agents = tools.Map(tools.Delegate(IsAffiliatedWith), tools.Filter(lambda x: 'awardeeName' in x, ctx))
+    related_agents = tools.Map(
+        tools.Delegate(IsAffiliatedWith),
+        tools.Filter(lambda x: 'awardeeName' in x, ctx)
+    )
 
+    identifiers = tools.Map(
+        tools.Delegate(AgentIdentifier),
+        tools.Try(
+            tools.IRI(ctx.piEmail),
+            exceptions=(ValueError,)
+        )
+    )
 
-class ContributorRelation(Parser):
-    schema = 'Contributor'
+    class Extra:
+        pi_last_name = ctx.piLastName
+        pi_first_name = ctx.piFirstName
+        pi_middle_initial = tools.Try(ctx.piMiddeInitial)
+        pi_email = tools.Try(ctx.piEmail)
 
-    agent = tools.Delegate(ContributorAgent, ctx)
-    cited_as = tools.Join(tools.Concat(ctx.piFirstName, ctx.piLastName), joiner=' ')
+
+class PIContributorRelation(Parser):
+    schema = 'PrincipalInvestigatorContact'
+
+    agent = tools.Delegate(PIContributorAgent, ctx)
+    cited_as = tools.Join(
+        tools.Concat(ctx.piFirstName, ctx.piLastName),
+        joiner=' '
+    )
 
 
 class AgentWorkRelation(Parser):
     agent = tools.Delegate(AffiliatedAgent, ctx)
 
 
 class CreativeWork(Parser):
+    # https://www.research.gov/common/webapi/awardapisearch-v1.htm#request-parameters
+
     title = ctx.title
+    description = ctx.abstractText
 
     identifiers = tools.Map(tools.Delegate(WorkIdentifier), ctx)
 
     related_agents = tools.Concat(
         tools.Map(tools.Delegate(FunderRelation), ctx),
-        tools.Map(tools.Delegate(ContributorRelation), ctx),
-        tools.Map(tools.Delegate(AgentWorkRelation), tools.Filter(lambda x: 'awardeeName' in x, ctx))
+        tools.Map(tools.Delegate(PIContributorRelation), ctx),
+        tools.Map(
+            tools.Delegate(POContributorRelation),
+            tools.Filter(lambda x: 'poName' in x, ctx)
```

Comment:
This seems to imply that some records have no program officer, in which case `poName` is omitted instead of set `null`. Is that right? (ditto awardeeName)

Reply:
It doesn't create a relation if `poName` isn't there. Is that the question?

Reply:
I think it will (try to?) create a relation if `poName` is null, though. If a field is irrelevant, do they omit the key or use null?

Reply:
They omit keys that aren't there. None of this is documented but I haven't seen any null values for any fields.

Reply:
I'd be more comfortable with something like `lambda x: x.get('poName')`, just in case that undocumented behavior changes.